core/policy.go

```diff
@@ -0,0 +1,88 @@
+package core
+
+import (
+	"context"
+
+	logging "gx/ipfs/QmbkT7eMTyXfpeyB3ZMxxcxg7XH8t6uXp49jqzz4HB7BGF/go-log"
+
+	"github.com/filecoin-project/go-filecoin/address"
+	"github.com/filecoin-project/go-filecoin/chain"
+	"github.com/filecoin-project/go-filecoin/types"
+)
+
+var log = logging.Logger("mqueue")
+
+// The outbound queue object on which the policy acts.
+type policyTarget interface {
+	RemoveNext(sender address.Address, expectedNonce uint64) (msg *types.SignedMessage, found bool, err error)
+	ExpireBefore(stamp uint64) map[address.Address][]*types.SignedMessage
+}
+
+// MessageQueuePolicy manages a target message queue state in response to changes on the blockchain.
+// Messages are removed from the queue as soon as they appear in a block that's part of a heaviest chain.
+// At this point, messages are highly likely to be valid and known to a large number of nodes,
+// even if the block ends up as an abandoned fork.
+// There is no special handling for re-orgs and messages do not revert to the queue if the block
+// ends up childless (in contrast to the message pool).
+type MessageQueuePolicy struct {
+	// The queue on which this policy acts
+	queue policyTarget
+	store chain.BlockProvider
+	// Maximum difference in message stamp from current block height before expiring an address's queue
+	maxAgeRounds uint64
+}
+
+// NewMessageQueuePolicy returns a new policy which removes mined messages from the queue and expires
+// messages older than `maxAgeRounds` rounds.
+func NewMessageQueuePolicy(queue *MessageQueue, store chain.BlockProvider, maxAge uint64) *MessageQueuePolicy {
+	return &MessageQueuePolicy{queue, store, maxAge}
+}
+
+// OnNewHeadTipset updates the policy target in response to a new head tipset.
+func (p *MessageQueuePolicy) OnNewHeadTipset(ctx context.Context, oldHead, newHead types.TipSet) error {
+	_, newBlocks, err := CollectBlocksToCommonAncestor(ctx, p.store, oldHead, newHead)
+	if err != nil {
+		return err
+	}
+
+	// Remove from the queue all messages that have now been mined in new blocks.
+
+	// Rearrange the blocks in increasing height order so messages are discovered in order.
```

Comment:
> We expect the network to only give us blocks with our message nonces strictly sequential. If somehow it doesn't, we'll error
Yes, something is wrong with either nonce calculation or chain sync and we want to know about it pretty loudly. I agree this will wedge message sending (which could be unwedged with a `Clear()`)

> You can also hit this if you restart and send a messages while other messages are waiting to be mined
The queue is in-memory only so must be empty if you restart. If you send a message while another, unknown to the node, is in progress of mining, the node will receive a message with the same nonce but different content, and successfully remove it and log an error (but not be stuck).